src/python/library/tritonclient/utils/cuda_shared_memory/__init__.py

```diff
@@ -258,9 +278,54 @@ def get_contents_as_numpy(cuda_shm_handle, datatype, shape):
             result = np.reshape(val, shape)
 
     finally:
-        c_int(_cshm_cuda_shared_memory_release_buffer(shm_addr))
+        c_int(_cshm_cuda_shared_memory_release_host_buffer(host_addr))
         return result
 
+def set_shared_memory_region_from_dlpack(cuda_shm_handle, input_values):
+    # this function basically is an implementation of 'from_dlpack'
```

Comment:
I understand the logic behind `set_shared_memory_region_from_dlpack`, but should we follow DLPack standard and think about the addition of `from_dlpack`, which people are used to?

Reply:
Yes I would like to have a `from_dlpack` function. But I think we can discuss on how to present it first, below is my initial thoughts and it's open for debate:

Different from regular `from_dlpack` where the function will allocate memory for holding the tensor, the shared memory use case often pre-allocates a memory buffer (shared memory region) and re-use it for different inferences. To follow this pattern, `from_dlpack` should be a class function of the shared memory region abstraction instead of a global function within the Python module, i.e.:
```
class SharedMemoryRegion:
    def from_dlpack(self, external_tensor, region_offset=None) -> SharedMemoryTensor
        # also need to finalize 'region_offset' detail and how it is
        # preserved / reset across invocation. Imagine the user may
        # call `from_dlpack` of the same region multiple times to
        # set all inputs for an inference.
```
and the returned `SharedMemoryTensor` will be the view of the region buffer that was written.

Ideally `SharedMemoryTensor` will become the replacement of the region-specific arguments passed into `InferInput.set_shared_memory`.

Note that this requires refactor on shared memory module to be more object oriented, the current implementation is a mirror of the underlying C library which is not consistent with the rest of the client code. The main refactor goal is to provide a class wrapper of `shm_handle`.